feat :  table & tableRow

jiji-hoon96 opened the pull request:
# Description of Changes

---
## Support 'table' and 'table-row' blocks
- Added Table and TableRow components to render table structures from Notion data.
- Implemented custom rendering logic for tables with column and row headers.

## Review point
- Should we integrate an existing table rendering solution or use our custom implementation for Notion tables?
- Is our current approach of handling column and row headers in the Table component sufficient and flexible?

## To reproduce
- npm run story:start
- Click 'Table' on the Storybook left side panel

## Screenshot
<img width="1450" alt="스크린샷 2024-09-27 오후 3 13 13" src="https://github.com/user-attachments/assets/a1f3d8a3-863d-4c61-8e75-ce50e7a458ec">

## Review Guide
Reviews are conducted based on priority levels, such as p0, p1, p2, p3, p4, and p5.  
p0 ~ p2: If the author decides not to reflect a review for p0 to p2, it signals that a proper discussion with the reviewer is  
necessary. It is expected that the review will be resolved either through incorporating the feedback or through further
 discussion.  
p3: indicates that the reviewer has identified a significant issue, but either lacks a clear solution or the comment lacks sufficient context. Further explanation or additional discussion on the reviewer's concerns is needed.
p4, p5: p4 and p5 suggest low priority, and if the author does not deem them important, these comments can be disregarded.

Discussion points for table and table-row implementation:

1. Table Component (p1):
   - We've implemented a custom Table component that handles both column and row headers based on the Notion JSON structure.
   - The component uses the `has_column_header` and `has_row_header` flags to determine the table structure.
   - Question: Is this approach flexible enough to handle various table configurations from Notion?

2. TableRow Component (p2):
   - The TableRow component renders each row of the table, including the header row.
   - It doesn't distinguish between header and regular rows, relying on the parent Table component for structure.
   - Question: Should we add more logic to TableRow to handle different types of rows (header vs. data)?

3. Styling (p3):
   - Currently, we're using basic CSS classes for styling. 
   - Question: Should we implement more advanced styling options, perhaps using a CSS-in-JS solution for more dynamic styling based on Notion's color and formatting options?

4. Performance Considerations (p2):
   - For large tables, rendering might become a performance issue.
   - Question: Should we implement virtualization for large tables, or is this out of scope for the current implementation?

5. Accessibility (p0)

6. Testing (p0)

Moon-DaeSeung commented:
Thank you for your excellent work on the Table and TableRow components! Your attention to detail and semantic HTML structure is commendable. Here's my review with some suggestions for improvement:
P2: Notion-like Styling
While your implementation is clean, we need to align more closely with Notion's design:

Please refer to Notion's styling for header cells and border colors. Our goal is to replicate Notion's UI as accurately as possible.
I recommend using CSS variables for cell and border colors, as our existing Notion components already consider dark mode. You can reference the top of our index.css file for examples.
The current implementation includes background colors for even rows, which isn't part of Notion's default style. Let's remove this for now to maintain consistency with Notion's design.

Remember, we're creating base components that mimic Notion's UI. In the future, we plan to offer customized theme versions of these components, which will allow for more diverse designs.
P4: Column Width Adjustment
Unfortunately, Notion's API doesn't provide width information for individual columns, so adjustments made in Notion aren't reflected in our implementation. To address this, we're planning to add an interaction allowing users to adjust column widths by dragging. However, this feature isn't required for the current PR and can be implemented in a future iteration.
Additional Notes:
Regarding large table performance: This is an excellent point to consider, but it's more relevant for our future database component rather than this table component. I'll create an issue to track this for future development.
Your use of semantic HTML tags for the table structure is excellent and greatly appreciated!
Great work on this PR. These suggestions aim to improve design accuracy and set the stage for future enhancements. Please focus on the P2 items for this PR. Let me know if you have any questions or need clarification on any point.

jiji-hoon96 commented:
New Requirement Added to Existing PR @Moon-DaeSeung 

I've implemented a new requirement in the existing pull request #40. I'd appreciate if someone could review the changes and confirm that they meet the new requirements.

Enhanced Table and TableRow components to more closely match Notion's table design and functionality.

- Updated Table component to handle various table widths and column ratios
- Modified TableRow to support different cell styles (bold, italic, colors)
- Implemented support for column and row headers
- Added ability to handle links within table cells
- Improved type definitions to better match Notion API response 

Please let me know if you need any additional information or if you have any questions.

Moon-DaeSeung commented on packages/react-notion-custom/src/lib/index.css:
```diff
@@ -704,6 +704,59 @@
+  font-size: 14px;
+  line-height: 20px;
+  color: var(--notion-table-cell-color);
+  border-top: 1px solid var(--notion-table-border-color);
```
I suggest you --fg-color-5 (that aleady exists)

Moon-DaeSeung commented on packages/react-notion-custom/src/lib/index.css:
```diff
@@ -704,6 +704,59 @@
+
+[data-theme="light"] .notion,
+.notion {
+  --notion-table-border-color: rgba(55, 53, 47, 0.09);
```
I suggest you --fg-color-5 (that aleady exists)

Moon-DaeSeung commented on packages/react-notion-custom/src/lib/index.css:
```diff
@@ -704,6 +704,59 @@
+
+[data-theme="dark"] .notion {
+  --notion-table-border-color: rgba(255, 255, 255, 0.13);
+  --notion-table-header-background: rgba(255, 255, 255, 0.04);
```
I suggest your --bg-color-1 (that already exists

Moon-DaeSeung commented:
you'd better check this repo in order to make notion style compoenent

https://github.com/cozy-blog/svelte-notion-page/blob/temp/packages/svelte-notion-page/src/lib/default-notion-components/Table.svelte

Moon-DaeSeung commented:
Hi @jiji-hoon96 ,
Just checking in on the status of this PR. Have you had a chance to look at the requested changes? Let me know if you need any clarification or if there's anything holding you back.
Thanks!

Moon-DaeSeung commented:
Thank you for your effort in submitting the PR. After careful consideration, I’ve decided to close this PR as the current implementation doesn’t fully meet the project’s needs. I believe it would be best to give another contributor a chance to tackle this issue. I truly appreciate your time and understanding, and I hope we can collaborate on other opportunities in the future.
